scipy/spatial/transform/_rotation.pyx

```diff
@@ -349,7 +349,118 @@ cdef double[:, :] _compute_euler_from_quat(
+        np.ndarray[double, ndim=1] n2, np.ndarray[double, ndim=1] n3,
+        bint extrinsic
+):
+    # The algorithm assumes extrinsic frame transformations. The algorithm
+    # in the paper is formulated for rotation quaternions, which are stored
+    # directly by Rotation.
+    # Adapt the algorithm for our case by reversing both axis sequence and
+    # angles for intrinsic rotations when needed
+
+    # intrinsic/extrinsic conversion helpers
+    cdef int angle_first, angle_third
+    if extrinsic:
+        angle_first = 0
+        angle_third = 2
+    else:
+        n1, n3 = n3, n1
+        angle_first = 2
+        angle_third = 0
+
+    cdef double[:] n_cross = _cross3(n1, n2)
+    cdef double lamb = atan2(_dot3(n3, n_cross), _dot3(n3, n1))
+
+    cdef int correct_set = False
+    if lamb < 0:
+        # alternative set of angles compatible with as_euler implementation
+        n2 = -n2
+        lamb = -lamb
+        n_cross[0] = -n_cross[0]
+        n_cross[1] = -n_cross[1]
+        n_cross[2] = -n_cross[2]
+        correct_set = True
+
+    cdef double[:] quat_lamb = np.array([
+            sin(lamb / 2) * n2[0],
+            sin(lamb / 2) * n2[1],
+            sin(lamb / 2) * n2[2],
+            cos(lamb / 2)]
+    )
+
+    cdef Py_ssize_t num_rotations = quat.shape[0]
+
+    # some forward definitions
+    cdef double[:, :] angles = _empty2(num_rotations, 3)
+    cdef double[:] _angles # accessor for each rotation
+    cdef double[:] quat_transformed = _empty1(4)
+    cdef double a, b, c, d
+    cdef double half_sum, half_diff
+    cdef double eps = 1e-7
+    cdef int case
+
+    for ind in range(num_rotations):
+        _angles = angles[ind, :]
+        _compose_quat_single(quat_lamb, quat[ind, :], quat_transformed)
+
+        # Step 1
+        # Permutate quaternion elements
+        a = quat_transformed[3]
+        b = _dot3(quat_transformed[:3], n1)
+        c = _dot3(quat_transformed[:3], n2)
+        d = _dot3(quat_transformed[:3], n_cross)
+
+        # Step 2
+        # Compute second angle...
+        _angles[1] = 2 * atan2(hypot(c, d), hypot(a, b))
+
+        # ... and check if equal to is 0 or pi, causing a singularity
+        if abs(_angles[1]) <= eps:
+            case = 1
+        elif abs(_angles[1] - <double>pi) <= eps:
+            case = 2
+        else:
+            case = 0 # normal case
+
+        # Step 3
+        # compute first and third angles, according to case
+        half_sum = atan2(b, a)
+        half_diff = atan2(d, c)
+
+        if case == 0:  # no singularities
+            _angles[angle_first] = half_sum - half_diff
+            _angles[angle_third] = half_sum + half_diff
+
+        else:  # any degenerate case
+            _angles[2] = 0
+            if case == 1:
+                _angles[0] = 2 * half_sum
+            else:
+                _angles[0] = 2 * half_diff * (-1 if extrinsic else 1)
+
+        _angles[1] -= lamb
+
+        if correct_set:
+            _angles[1] = -_angles[1]
```

Comment:
Can this block be refactored to be used in both this function and `_compute_euler_from_quat`?

Reply:
I think it could, woudn't it introduce too many subfunctions though? Since transforming it to Euler angles already includes a call to `as_euler` which calls either `compute_euler_from_quat` or `compute_euler_from_matrix`

Reply:
I actually forgot that "from_matrix" still exists, I don't think it's particular important and can be removed at any point.

I don't think that nesting calling is a problem by itself. In this case it can emphasize that 2 methods are closely related and maybe save some lines of code.

Anyway, can you try it and see how it looks?

Reply:
Done, but I haven't tested if it changes something considerably for execution time.